libcontainerd/remote/client.go

```diff
@@ -200,10 +201,21 @@ func (c *container) Start(ctx context.Context, checkpointDir string, withStdin b
 	if runtime.GOOS != "windows" {
 		taskOpts = append(taskOpts, func(_ context.Context, _ *containerd.Client, info *containerd.TaskInfo) error {
 			if c.v2runcoptions != nil {
-				opts := *c.v2runcoptions
-				opts.IoUid = uint32(uid)
-				opts.IoGid = uint32(gid)
-				info.Options = &opts
+				// libcontainerd/remote/client.go:204:13: copylocks: assignment copies lock value to opts: github.com/docker/docker/vendor/github.com/containerd/containerd/runtime/v2/runc/options.Options contains github.com/docker/docker/vendor/google.golang.org/protobuf/internal/impl.MessageState contains sync.Mutex (govet)
+				//				opts := *c.v2runcoptions
+				//				        ^
+				info.Options = &v2runcoptions.Options{
+					NoPivotRoot:   c.v2runcoptions.GetNoPivotRoot(),
+					NoNewKeyring:  c.v2runcoptions.GetNoNewKeyring(),
+					ShimCgroup:    c.v2runcoptions.GetShimCgroup(),
+					IoUid:         uint32(uid),
+					IoGid:         uint32(gid),
+					BinaryName:    c.v2runcoptions.GetBinaryName(),
+					Root:          c.v2runcoptions.GetRoot(),
+					SystemdCgroup: c.v2runcoptions.GetSystemdCgroup(),
+					CriuImagePath: c.v2runcoptions.GetCriuImagePath(),
+					CriuWorkPath:  c.v2runcoptions.GetCriuWorkPath(),
+				}
```

Comment:
Does anyone have a good approach for these kind of things? So the issue is that these types embed a mutex (through the `state` (`protoimpl.MessageState`); the linter (correctly) flags the old code because we're copying the mutex. https://github.com/containerd/containerd/blob/v1.7.0-beta.0/runtime/v2/runc/options/oci.pb.go#L23-L31

```go
type Options struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	// disable pivot root when creating a container
	NoPivotRoot bool `protobuf:"varint,1,opt,name=no_pivot_root,json=noPivotRoot,proto3" json:"no_pivot_root,omitempty"`
	// create a new keyring for the container
	NoNewKeyring bool `protobuf:"varint,2,opt,name=no_new_keyring,json=noNewKeyring,proto3" json:"no_new_keyring,omitempty"`
```


The reason we copy is that we want to have the existing options, but update some fields; So my naive approach is to construct a new option, and copy all the fields. While that works; it's brittle, as (e.g.) if a field is added or removed, this code will not copy those new fields.

I wish these types had a `.Copy()` method to get a copy of them (to prevent the mutex being copied), but perhaps there's a better approach for this.


Related discussion on StackOverflow; https://stackoverflow.com/a/64183968/1811501, which seems to indicate this is _on purpose_, and the `MessageState` has a `pragma.DoNotCopy
` (and in this case `pragma.DoNotCompare)` to indicate it **should not be copied**; https://github.com/protocolbuffers/protobuf-go/blob/v1.28.1/internal/impl/message_reflect.go#L358-L364

```go
type MessageState struct {
	pragma.NoUnkeyedLiterals
	pragma.DoNotCompare
	pragma.DoNotCopy
```

So, perhaps we're doing it wrong? Or should the type be changed?

Reply:
Maybe I'm looking at [`proto.Clone()`](https://pkg.go.dev/google.golang.org/protobuf@v1.28.1/proto#Clone) or [`proto.Merge`](https://pkg.go.dev/google.golang.org/protobuf@v1.28.1/proto#Merge)

Reply:
^^ updated to use `proto.Clone()`